[Toh Yi Fan] iP

TY1Fan opened the pull request:
# Fanny

> "Chatbots will be your new best friend." - Christine Crandell ([Forbes](https://www.forbes.com/sites/christinecrandell/2016/10/23/chatbots-will-be-your-new-best-friend/))

#### Fanny frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

#### All you need to do is,
1. download it from [here](https://github.com/TY1Fan/ip/releases).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😆 

And it is **FREE!**

#### Features:

- [x] Managing todo list
- [x] Managing deadlines
- [x] Managing events
- [ ] Reminder

##
If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```
public static void main(String[] args) {
    new Fanny("./data/fanny.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

TY1Fan committed:
Rename main class

TY1Fan committed:
Add greeting

TY1Fan committed:
Add ability to echo to commands

TY1Fan committed:
Add support for list

TY1Fan committed:
Fix bug in list

TY1Fan committed:
Fix bug for not closing scanner and initialization of scanner

TY1Fan committed:
Refactor code to adhere to OOP and style guide

TY1Fan committed:
Add task class

TY1Fan committed:
Add support for marking of task as done

TY1Fan committed:
Add support for unmarking of task

TY1Fan committed:
Fix bug for unintended side-effect of creating additional empty items

TY1Fan committed:
Add support for tracking todo tasks

TY1Fan committed:
Add support for tracking deadline

TY1Fan committed:
Add support for tracking of event

TY1Fan committed:
Add automated text ui testing

TY1Fan committed:
Refactor code and add exception handling

TY1Fan committed:
Add support for deleting task and fix bug regarding exceptions

TY1Fan committed:
Add support for file operations and saving of data

TY1Fan committed:
Revert "Add support for file operations and saving of data"

This reverts commit d68443c8d72fe6ea39d8ac55dc8a8a64f9ac55f5.

TY1Fan committed:
Revert "Revert "Add support for file operations and saving of data""

This reverts commit 89d15ce8cb69ffc6a05c5892fc3fd47723513fdb.

TY1Fan committed:
Merge branch 'branch-Level-7'

TY1Fan committed:
Add support for deadlines date as java.time.LocalDateTime

TY1Fan committed:
Add support for storing event date and time as java.time.LocalDateTime

TY1Fan committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Add support for storing event date and time as java.time.LocalDateTime
  Add support for deadlines date as java.time.LocalDateTime

TY1Fan committed:
Refactor code for Storage which handles file operations

TY1Fan committed:
Refactor code to include Ui which deals with interactions with the user

TY1Fan committed:
Refactor code to include parser and commands which deal with making sense of the user command

TY1Fan committed:
Merge branch 'branch-A-MoreOOP'

* branch-A-MoreOOP:
  Refactor code to include parser and commands which deal with making sense of the user command
  Refactor code to include Ui which deals with interactions with the user
  Refactor code for Storage which handles file operations

TY1Fan committed:
Organize the classes into suitable java packages

TY1Fan committed:
Merge branch 'branch-A-Packages'

* branch-A-Packages:
  Organize the classes into suitable java packages

TY1Fan committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

TY1Fan committed:
Add JUnit tests to test behavior of Event and Deadline classes

TY1Fan committed:
Merge branch 'branch-A-JUnit'

* branch-A-JUnit:
  Add JUnit tests to test behavior of Event and Deadline classes

TY1Fan committed:
Change build.gradle file

TY1Fan committed:
Add Javadoc comments to classes in task package

TY1Fan committed:
Add Javadoc comments for Ui.java

TY1Fan committed:
Add javadoc for storage.java

TY1Fan committed:
Add javadoc comments for parse.java

TY1Fan committed:
Add javadocs comment for command packaged classes

TY1Fan committed:
Configure intellij to not indent case statements

TY1Fan committed:
Add support for finding of tasks

TY1Fan committed:
Add javadoc comments for the find increment

TY1Fan committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Configure intellij to not indent case statements

TY1Fan committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add javadoc comments for the find increment
  Add support for finding of tasks

TY1Fan committed:
Restore find files and javadocs for find

TY1Fan committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Restore find files and javadocs for find

# Conflicts:
#	src/main/java/fanny/parser/Parser.java
#	src/main/java/fanny/task/TaskList.java

Resolution: keep both

TY1Fan committed:
Add support for checkstyle using gradle

TY1Fan committed:
Fix style issues using checkstyle

TY1Fan committed:
Merge branch 'branch-A-CheckStyle'

* branch-A-CheckStyle:
  Fix style issues using checkstyle
  Add support for checkstyle using gradle

TY1Fan committed:
Update build.gradle for javafx

TY1Fan committed:
Set up javafx program

TY1Fan committed:
Design layout of gui

TY1Fan committed:
Refactor code to use fxml

TY1Fan committed:
Integrate past code with new gui

TY1Fan committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Integrate past code with new gui
  Refactor code to use fxml
  Design layout of gui
  Set up javafx program
  Update build.gradle for javafx

benjiBase reviewed (commented):
LGTM. Nice work

IanCheah commented on src/main/java/fanny/Fanny.java:
```diff
@@ -0,0 +1,105 @@
+     */
+    public void run() {
+        ui.printHello();
+        boolean isExit = false;
```
I think that isExit does not sound very boolean 😅 . I have a few suggestions. Perhaps you can use something like "hasExited", "shouldExit", or "isFinshed".

TY1Fan replied:
Thank you for your comment. I believe shouldExit is indeed more apt in this case as the method specifies whether the program should exit after executing the command.

IanCheah reviewed (commented):
LGTM. Nice work.

IanCheah commented on src/main/java/fanny/parser/Parser.java:
```diff
@@ -0,0 +1,56 @@
+        String action = cmdParts[0];
+        String arguments = cmdParts.length > 1 ? cmdParts[1] : "";
+
+        switch (action) {
```
Good formatting of switch 😄 . Common mistake would be to indent the case.

TY1Fan replied:
Thank you!

IanCheah commented on src/main/java/fanny/storage/Storage.java:
```diff
@@ -0,0 +1,140 @@
+     * @throws FileNotFoundException If the file does not exist.
+     */
+    public List<Task> loadTask() throws FileNotFoundException {
+        List<Task> tasks = new ArrayList<>();
```
Good. Plural for tasks.

TY1Fan replied:
Thank you!

IanCheah commented on src/main/java/fanny/Fanny.java:
```diff
@@ -0,0 +1,105 @@
+
+            String capturedOutput = outputStream.toString();
+
+            String filteredOutput = Arrays.stream(capturedOutput.split(System.lineSeparator()))
```
Good indentation of 8 spaces. 😄

TY1Fan replied:
Thank you!

IanCheah commented on src/main/java/fanny/Fanny.java:
```diff
@@ -0,0 +1,105 @@
+package fanny;
+
+import java.io.ByteArrayOutputStream;
```
Good. No wildcard imports. 😄

TY1Fan replied:
Thank you!

IanCheah reviewed (commented):
LGTM. nice work bro

Jollybomber commented on src/main/java/fanny/Fanny.java:
```diff
@@ -0,0 +1,105 @@
+    /**
+     * Generates a response for the user's chat message.
+     */
+    public String getResponse(String input) {
```
getResponse sounds like we are getting a user response, perhaps generateResponse?

TY1Fan replied:
Thank you for your comment. I think generateResponse is indeed more appropriate.

Jollybomber commented on src/main/java/fanny/FannyException.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Represents an exception unique to the Fanny application.
+ */
+public class FannyException extends Exception {
```
Good naming of class to specify Exception that is specific to Fanny! 😃

TY1Fan replied:
Thank you!

Jollybomber commented on src/main/java/fanny/command/Command.java:
```diff
@@ -0,0 +1,34 @@
+     * @param list The task list to be modified by the command.
+     * @param ui The user interface to interact with the user.
+     */
+    public abstract void actionable(TaskList list, Ui ui);
```
Perhaps the naming could provide more detail about the functionality of this method?

Jollybomber commented on src/main/java/fanny/task/Event.java:
```diff
@@ -0,0 +1,59 @@
+public class Event extends Task {
+
+    /** The start time of the event. */
+    protected LocalDateTime from;
```
Perhaps startDate/startTime and endDate/endTime would be more accurate?

TY1Fan replied:
Thank you for your comment. I think startTime and endTime is indeed more accurate.

Jollybomber reviewed (commented):
Naming generally follows style guide suggested by the textbook. Good job! 😄

TY1Fan committed:
Generate a new jar file

TY1Fan committed:
Add assertions for tasklist

Some methods in tasklist contains assumptions:
* markAsDone assumes that index provided is valid
* markAsNotDone assumes that index provided is valid
* findTasks assumes that there is a keyword to search for

Let's add Java assertions to verify these assumptions at runtime.

TY1Fan committed:
Refactor files in command package

Some files in the command package contain code quality violations:
* poor naming such as actionable and isExit
* some method implementations are also too long
* multiple layers of abstraction within a code fragment

So let's:
* use names that is more self-explanatory
* extract out smaller methods
* use methods to group abstractions

TY1Fan committed:
Refactor files in task package

Some files contain code quality violations:
* poor naming like to, from
* complicated expressions like chaining of variables

So let's:
* use names like startTime, endTime
* simplify expressions by variable assignment

TY1Fan committed:
Refactor storage class

Storage class contains code quality violations:
* some methods are too long
* some expressions are too complicated

So let's:
* refactor by extracting out smaller methods
* use more intermediate variables

TY1Fan committed:
Refactor gui components

Some gui files have code quality violations:
* some methods are too long
* multiple layers of abstraction

So let's:
* refactor by extracting reusable methods

TY1Fan committed:
Fix bug for MainWindow

TY1Fan committed:
Use streams to find tasks

TY1Fan committed:
Merge pull request #2 from TY1Fan/branch-A-Assertions

Branch-A-Assertions

TY1Fan committed:
Revert "Branch-A-Assertions"

TY1Fan committed:
Merge pull request #4 from TY1Fan/revert-2-branch-A-Assertions

Revert "Branch-A-Assertions"

TY1Fan committed:
Merge branch 'master' of https://github.com/TY1Fan/ip

* 'master' of https://github.com/TY1Fan/ip:
  Revert "Branch-A-Assertions"
  Add assertions for tasklist
  Generate a new jar file

TY1Fan committed:
Revert "Revert "Branch-A-Assertions""

TY1Fan committed:
Merge pull request #5 from TY1Fan/revert-4-revert-2-branch-A-Assertions

Revert "Revert "Branch-A-Assertions""

TY1Fan committed:
Merge branch 'master' of https://github.com/TY1Fan/ip

* 'master' of https://github.com/TY1Fan/ip:
  Revert "Revert "Branch-A-Assertions""

TY1Fan committed:
Merge branch 'master' into branch-A-CodeQuality

TY1Fan committed:
Merge pull request #1 from TY1Fan/branch-A-CodeQuality

Branch-A-Code-Quality

TY1Fan committed:
Merge branch 'master' of https://github.com/TY1Fan/ip

* 'master' of https://github.com/TY1Fan/ip:
  Fix bug for MainWindow
  Refactor gui components
  Refactor storage class
  Refactor files in task package
  Refactor files in command package

TY1Fan committed:
Merge pull request #3 from TY1Fan/branch-A-Streams

Branch-A-Streams

TY1Fan committed:
Merge branch 'master' of https://github.com/TY1Fan/ip

* 'master' of https://github.com/TY1Fan/ip:
  Use streams to find tasks

TY1Fan committed:
Add support for reminders

TY1Fan committed:
Merge branch 'branch-B-Reminders'

* branch-B-Reminders:
  Add support for reminders

TY1Fan committed:
Merge branch 'branch-A-CodeQuality' of https://github.com/TY1Fan/ip into branch-A-CodeQuality

* 'branch-A-CodeQuality' of https://github.com/TY1Fan/ip:
  Revert "Revert "Branch-A-Assertions""
  Revert "Branch-A-Assertions"
  Add assertions for tasklist
  Generate a new jar file

TY1Fan committed:
Merge pull request #6 from TY1Fan/branch-A-CodeQuality

Merge branch 'branch-A-CodeQuality' of https://github.com/TY1Fan/ip i…

TY1Fan committed:
Merge branch 'master' of https://github.com/TY1Fan/ip

* 'master' of https://github.com/TY1Fan/ip:

TY1Fan committed:
Refactor command package and fanny

TY1Fan committed:
Merge branch 'master' of https://github.com/TY1Fan/ip into branch-A-CheckStyle

* 'master' of https://github.com/TY1Fan/ip:
  Refactor command package and fanny
  Add support for reminders
  Revert "Revert "Branch-A-Assertions""
  Revert "Branch-A-Assertions"
  Use streams to find tasks
  Fix bug for MainWindow
  Refactor gui components
  Refactor storage class
  Refactor files in task package
  Refactor files in command package
  Add assertions for tasklist
  Generate a new jar file
  Integrate past code with new gui
  Refactor code to use fxml
  Design layout of gui
  Set up javafx program
  Update build.gradle for javafx

TY1Fan committed:
Update code for styling

TY1Fan committed:
Merge branch 'branch-A-CheckStyle'

* branch-A-CheckStyle:
  Update code for styling

TY1Fan committed:
Fix javadoc header comments

TY1Fan committed:
Set gui dimensions

TY1Fan committed:
Improve gui using css

TY1Fan committed:
Merge branch 'branch-A-BetterGui'

* branch-A-BetterGui:
  Improve gui using css
  Set gui dimensions

TY1Fan committed:
Change phrases for commands

TY1Fan committed:
Merge branch 'branch-A-Personality'

* branch-A-Personality:
  Change phrases for commands

TY1Fan committed:
Fix minor bugs

TY1Fan committed:
Update README.md

TY1Fan committed:
Merge branch 'branch-A-UserGuide'

* branch-A-UserGuide:
  Update README.md

TY1Fan committed:
Fix bug for exiting app

TY1Fan committed:
Fix bug for mark cmd

TY1Fan committed:
Add more error handling

TY1Fan committed:
Add error handling for event

TY1Fan committed:
Fix assertions and README

TY1Fan committed:
Fix javadocs and readme

TY1Fan committed:
Add AI.md

TY1Fan committed:
Merge branch 'branch-A-AiAssisted'

* branch-A-AiAssisted:
  Add AI.md

TY1Fan committed:
Fix bug for remind

TY1Fan committed:
Fix bugs in commands

TY1Fan committed:
Fix bug for storage

TY1Fan committed:
Fix bug for commands

TY1Fan committed:
Fix some bugs

TY1Fan committed:
Update readme

TY1Fan committed:
Fix bug for bye

TY1Fan committed:
Fix bug in readme
